lib/Doctrine/ORM/UnitOfWork.php

```diff
@@ -418,17 +418,39 @@ public function commit($entity = null)
 
         $this->dispatchPostFlushEvent();
 
+        $this->postCommitClear($entity);
+    }
+
+    /**
+     * @param null|object|array $entity
+     */
+    private function postCommitClear($entity = null) {
+
         // Clear up
         $this->entityInsertions =
         $this->entityUpdates =
         $this->entityDeletions =
         $this->extraUpdates =
-        $this->entityChangeSets =
         $this->collectionUpdates =
         $this->collectionDeletions =
         $this->visitedCollections =
-        $this->scheduledForSynchronization =
         $this->orphanRemovals = array();
+
+        if ($entity === null) {
+            $this->entityChangeSets = $this->scheduledForSynchronization = array();
+        } elseif (is_object($entity)) {
+            $oid = spl_object_hash($entity);
+            $class = $this->em->getClassMetadata(get_class($entity));
+            $this->clearEntityChangeSet($oid);
+            $this->clearScheduledForSynchronization($class, $oid);
+        } elseif (is_array($entity)) {
+            foreach ($entity as $object) {
```

Comment:
The block inside this `foreach` can be a private method reused both here and in the `is_object` block